### PR TITLE
Allow multiple texture cache entries for textures at the same address

### PR DESCRIPTION
**Pull request by mimimi085181**
This is the same trick which is used for Metroid's fonts/texts, but for all textures. If 2 different textures at the same address are loaded during the same frame, create a 2nd entry instead of overwriting the existing one. If the entry was overwritten in this case, there wouldn't be any caching, which results in a big performance drop.

The restriction to textures, which are loaded during the same frame, prevents creating lots of textures when textures are used in the regular way. This restriction is new. Overwriting textures, instead of creating new ones is faster, if the old ones are unlikely to be used again.

Since this would break efb copies, don't do it for efb copies.

Castlevania 3 goes from 80 fps to 115 fps for me.

There might be games that need a higher texture cache accuracy with this, but those games should also see a performance boost from this PR.

Some games, which use paletted textures, which are not efb copies, might be faster now. And also not require a higher texture cache accuracy anymore. (similar sitation as PR dolphin-emu#1916)

### CODE REVIEW

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by degasus** [comment]
I think it should be fine to remove the old tlut_hash dependency of the texID completely in this way. So both may be cached longer

**Comment by mimimi085181** [comment]
@degasus: I don't understand. If you remove mixing the texID with the tlut_hash, wouldn't this result in the same texture being overwritten for different palettes, over and over. I thought Metroid has some texture for fonts/texts that only differs in the palette. So it's faster when all those texture+palette combinations get their own texture cache entry, because those can be loaded again during the next frame.

But i think the trick for the paletted textures should be reduced to textures from the same frame too. Right now, if a game updates the palette of a texture, there's a new cache entry generated. If the old one would be overwritten, it would be faster, because it's likely to not be used again. Without the restriction to the same frame, New Super Mario Bros got around 10% slower for me. I think the same might happen to games with paletted textures, that are not used the same way as Metroid does, and which are not efb copies of course.

**Comment by mimimi085181** [comment]
Updated, so the entry with texID is reused, if the current hash matches the hash of the entry at texID. The 1st version of this PR would create a 2nd entry, although the existing entry was perfectly useable.

**Comment by Linktothepast** [comment]
So what kind of games would this cause a speedup, games that need the safest tex cache setting? Or is it mainly for Castlevania 3?

**Comment by mimimi085181** [comment]
It is written for Castlevania 3, but it might affect other games. It's for games that use the same memory area over and over for different objects. But i doubt there are many games that do this. Maybe some odd and simple game, that runs way worse than it should, compared to other games. Or another game with its own emulator.

There might be small regressions from this, mainly games that require a higher texture cache accuracy when using this PR. Having multiple versions of a texture in the cache, means that there are more chances for 2 textures to have the same hash, although they are different. I think that's what happened with the efb copies that are paletted textures.

For now the build bot only built the 1st commit, the 2nd commit possibly changes things for better or worse, so please only do tests after that is built.

**Comment by Linktothepast** [comment]
If it affects mainly one virtual console game and is also a possible cause of regressions can't say that it sounds exiting as a change...

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by mimimi085181** [comment]
I only see a possibility for said regression(needing higher cache accuracy) for games that use the same address for different textures(during the same frame). This PR is aiming at handling this situation faster, so it's likely that it's still faster even with a higher cache accuracy. Well, that is if you move from medium, 512 samples, to something like 4096 samples, instead of safe, which hashes the whole object.

As an example on how bad safe accuracy is:
This PR alone gets Castlevania 3 to around 115 fps for me, this PR + 4096 samples, instead of safe accuracy, gets it to around 210 fps. Master gets only to around 80 fps. 4096 samples alone was around 130 i think.

All in all, i hope this fixes the performance of some odd edge cases, which perform really badly for no apparent reason. Like Castlevania 3.

**Comment by Linktothepast** [comment]
Here it seems that
 [Video_Settings]
SafeTextureCacheColorSamples = 2048
should be sufficient with master, could you confirm from your side? And it goes from 90 to 150 fps. Not bad imo and should cause no problems with other games.

**Comment by mimimi085181** [comment]
I wrote a small test problem, which logs which sample number is required for which texture. I think it said 4096 during the 1st stage in castlevania 3.

I'm going to tweak that test build a little and then post it on the developer sub forum or something.

**Comment by magumagu** [comment]
If we want to allow multiple textures with the same address, please just make TextureCache::textures an std::multimap.  That would make the intent a lot more clear.

**Comment by mimimi085181** [comment]
Thanks, i'll look into that. But i only want to use multiple entries for the same address under strict conditions, because otherwise, there are regressions for many games.

**Comment by mimimi085181** [comment]
Please don't review the comments in the code. Those need a big update!

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 1]
```diff
@@ -344,16 +344,39 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+	// Allow multiple cache entries for textures at the same address. This is a trick to make the Castlevania 3 faster, which
+	// uses the same address for different textures. Do not do this for paletted textures, because there's a similar trick used
+	// for them already. And do not do this for efb copies, because those would break with this. To prevent a regression caused
+	// by constantly creating new textures instead of overwriting old ones, only do this for textures that are loaded at the
+	// same frame.
+	TexCache::iterator iter = textures.lower_bound(texID);
+	if (iter != textures.end())
+	{
+		if (iter->second->IsEfbCopy() || (iter->second->frameCount != FRAMECOUNT_INVALID && textures.count(texID) == 1))
+		{
```
Comment (magumagu) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: The IsEfbCopy() check doesn't work correctly if textures.lower_bound doesn't find an element equal to texID.

Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Oh, i thought textures.lower_bound(texID) == textures.end(), if there's no entry with texID.

Comment (magumagu) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: No, if there's no entry with texID, textures.lower_bound(texID) == textures.upper_bound(texID).

Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Thanks, now i understand. Is there a faster/better way than:
if (iter != textures.end() && iter != textures.upper_bound(texID))
?
Hmm, i could do this:
if (textures.count(texID) != 0)
{
TexCache::iterator iter = textures.lower_bound(texID);
if (iter->second->IsEfbCopy() [...]

I think everytime i use lower_bound, upper_bound or count, the whole map is searched, so i should try to avoid calling those more ofthen than neccessary.

Comment (magumagu) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: You could use multimap::equal_range; you shouldn't need to use any other function on the map after that.

**Comment by mimimi085181** [comment]
Updated to use textures.equal_range as per magumagu's suggestion.

Also updated the comments in the code.

From my side this PR is ready now. But i think it's not clear yet whether this should be merged at all or not. Some comments on this topic are welcome.

My arguments:
+ Master currently uses the same trick on all paletted textures which are not efb copies, this PR uses it on less textures. So in some cases this PR might fix regressions caused by the trick for Metroid's fonts.
+ Castlevania 3 is faster, and hopefully some other yet unknown games.
o I only see a chance for a regression, if the same happens as it did with the paletted textures that are efb copies. Because there are more textures in the cache, there are more chances for a false positive match, when comparing hashes. But this will only happen on textures which benefit from this change, so if it happens, it might be +/- 0 compared to master.
- The code is more complicated than before

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 2]
```diff
@@ -329,31 +329,47 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+	TexCache::iterator iter = iter_range.first;
+	if (iter != iter_range.second)
+	{
+		if (iter->second->IsEfbCopy() || (iter->second->frameCount != FRAMECOUNT_INVALID && std::next(iter, 1) == iter_range.second))
+		{
+			entry = iter->second;
+		}
+		else
+		{
+			while (!entry && iter != iter_range.second)
```
Comment (magumagu) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: If I'm understanding this correctly, this is essentially an alternate mode: if we detect a game using multiple textures with the same address, we allow any number of textures with the same address.  This seems like it could easily blow up: if the heuristic gets triggered for one frame, it's essentially turned on permanently, which could kill performance.

Maybe just make Cleanup() a lot more aggressive in this case?

Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Hmm, i think we need to handle 2 different cases here:

Case 1:
If a game starts to update a texture once per frame, after it was using the same one for various things during one frame, this would blow up to max TEXTURE_KILL_THRESHOLD(currently 200) entries for one texture. This is exactly what i was trying to prevent with the limit to textures from the same frame. Something like this caused a 10% slowdown + rubber banding performance on New Super Mario Bros.

I think this change in beavior should be detectable. But it would complicate the logic quite a bit:
If there's no entry which was used duing the the same frame AND the highest frame number of all entries occurs exactly once, then switch to the update texture once per frame mode and overwrite the oldest entry. After TEXTURE_KILL_THRESHOLD frames, everything will be back to normal.

Case 2:
If a texure is used for 20 different things, and none of the old versions match the new ones, this could blow up to TEXTURE_KILL_THRESHOLD times 20, so 4000 entries. This would be bad, really bad. So should i put in a hard limit of allowed textures per texID? And if this is reached, overwrite the oldest entry, instead of creating a new one?

This may sound bad, but i would like to have a high limit for this. Something like 100 entries. This would only affect games that use the same texture over and over for various things. So it's possible, that this could be all textures used by the game or at least a lot. If the limit isn't high enough, the texture cache would fail as much as when just using one entry.

I need to test how many entries Metroid and Castlevania 3 generate here for the same texID. Hopefully this will give us a clue on a good limit.

**Comment by mimimi085181** [comment]
Updated the PR with a maximum number of texture cache entries for the each address. For Castlevania 3, a max of 2 would be sufficient, because it seems to "just" alternate between 2 different textures. But in Metroid Prime 3, i have seen up to 64 entries at once. I'll do some further tests on Metroid tomorrow, to see if i can set the max lower, without loosing performance.

I have given up on trying to find a clean way to detect a change in the use pattern from multiple textures at the same address -> normal texture update. This complicates the logic quite a bit and it's unlikely to happen. And if it does happen, the performance drop caused by this would be minimal with a small enough max number of entries.

**Comment by magumagu** [comment]
Setting a max number of textures with the same address seems fine.  Fifty seems a little on the high side given that the examples I've seen involve two or three textures.

**Comment by Stevoisiak** [comment]
@dolphin-emu-bot rebuild

**Comment by mimimi085181** [comment]
@magumagu: I have seen Metroid Prime 3 use around 15 textures at the same address at the same time. Setting the limit to 10 did not cause a slowdown there, but to be safe, i set the limit to 20. I want to be sure that this does not cause a regression on Metroid.

**Comment by Tilka** [comment]
@dolphin-emu-bot rebuild

**Comment by magumagu** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 3]
```diff
@@ -329,31 +330,76 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
 	}
 
 	// GPUs don't like when the specified mipmap count would require more than one 1x1-sized LOD in the mipmap chain
 	// e.g. 64x64 with 7 LODs would have the mipmap chain 64x64,32x32,16x16,8x8,4x4,2x2,1x1,0x0, so we limit the mipmap count to 6 there
 	tex_levels = std::min<u32>(IntLog2(std::max(width, height)) + 1, tex_levels);
 
-	TCacheEntryBase*& entry = textures[texID];
+	TCacheEntryBase* entry = nullptr;
+
+	// Find all texture cache entries for the current texture address, and decide whether to use one of them, or to create a new one
```
Comment (magumagu) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Minor nitpick: please wrap comments at 80 columns or so.

**Comment by magumagu** [comment]
Please rebase this for testing; this is currently based on master without #2005, which is a little awkward for testing a texcache change.

**Comment by mimimi085181** [comment]
@magumagu: Thanks to a lot of help on irc, this is now based on latest master. And i shortened the comment lines a bit. If they are still too long, i can make them shorter.

While testing i noticed that Rogue Squadron 3 is using the same texture address several times during the same frame. I did not notice any change in performance though, neither did 2 people on the forums after i asked them to check for regressions with the PR and RS3.

**Comment by mimimi085181** [comment]
Ok, now i'm confident there won't be a regression in Metroid. It's allowing more than the max number of entries now, if all the textures are used in the same frame.

So if Metroid was using 30 texture+palette combinations duing the same frame, the limit would go up to 30 automatically.

**Comment by magumagu** [comment]
@dolphin-emu-bot rebuild

I'm not completely happy that the cache can grow to an arbitrarily large size, but I guess that isn't really new.

**Comment by mimimi085181** [comment]
@magumagu: If there are let's say different 50 textures at the same address, used during the same frame, it's about choosing the lesser evil i think:
- Have 1 texture cache entry, which gets overwritten over and over again. With 20 entries, the best case would be that only the 1st entry is overwritten over and over, 30/31 times, and 19 entries stay in the cache.
- Have 50 textures in the cache, but if those textures get used in the next frame, they're cached.

**Comment by magumagu** [comment]
Either of those possibilities is probably okay... I'm more worried about the case where we have an hwtest that modifies a texture thousands of times without ever drawing a frame.  But I guess if we wanted to guard against that, we would really want an overall cap on the size of the cache rather than messing with the behavior here.

**Comment by mimimi085181** [comment]
Well, i'm still open for suggestions how to make this better. I just want Castlevania 3 and the Metroids to run as fast as possible, and this with generic code that doesn't know which game is running.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 4]
```diff
@@ -329,31 +330,81 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+	if (iter != iter_range.second)
+	{
+		if (iter->second->IsEfbCopy() || (iter->second->frameCount != FRAMECOUNT_INVALID && std::next(iter, 1) == iter_range.second))
+		{
+			entry = iter->second;
+		}
+		else
+		{
+			u32 counter = 0;
+			TexCache::iterator oldest_entry = iter;
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: c++11 multimap is defined to store the elements with the same ID in order, so iter is already the oldest texture.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 5]
```diff
@@ -24,6 +24,7 @@ static const u64 TEXHASH_INVALID = 0;
 static const int TEXTURE_KILL_THRESHOLD = 200;
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: You likely also want to reduce this one. I guess 200 was only used at the time when we also remove EFB copys. I guess so ~10 frames should be fine.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 6]
```diff
@@ -329,31 +330,81 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
-		//       we must make sure that a paletted texture gets assigned multiple IDs for each tlut used.
-		//
-		//       EFB copys however didn't know anything about the tlut, so don't change the texID if there
-		//       already is an efb copy at this source. This makes those textures less broken when using efb to texture.
-		//       Examples are the mini map in Twilight Princess and objects on the targetting computer in Rogue Squadron 2(RS2).
-		// TODO: Convert those textures using the right palette, so they display correctly
-		auto iter = textures.find(texID);
-		if (iter == textures.end() || !iter->second->IsEfbCopy())
-			texID ^= ((u32)tlut_hash) ^(u32)(tlut_hash >> 32);
+		// TODO: Convert paletted textures, which are efb copies, using the right palette, so they display correctly
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I think #2059 is a better place for this TODO ;)

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 7]
```diff
@@ -329,31 +330,81 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+		}
+		else
+		{
+			u32 counter = 0;
+			TexCache::iterator oldest_entry = iter;
+			int temp_frameCount = 0x7fffffff;
+
+			while (!entry && iter != iter_range.second)
+			{
+				if (tex_hash == iter->second->hash)
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I think just checking for the hash isn't enough. We likely just want to inline the next if(entry)..... condition here. Of course with the check to only use efb2tex copies if there are only one texture within the cache.

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 4 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: But i'd like to replace the entry, which hasn't been used for the longest time. Just replacing the 1st entry could result in removing and adding the same entry over and over again.

**Comment by mimimi085181** [comment]
Rebased, updated comments and merged the code as per degasus' suggestion. Thanks, the code is so much cleaner now.

This is now also trying to keep the cache as small as possible. If there are X entries used at the same frame, the cache blows up to X entries, but not any further. There might be regressions due to this, but allowing a too large number of entries is likely to cause problems on low end systems. In my opinion, the way it is done now, is the right balance.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 8]
```diff
@@ -329,56 +328,92 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+					temp_frameCount = iter->second->frameCount;
+					oldest_entry = iter;
+				}
+				++iter;
+			}
 
-		// pool this texture and make a new one later
-		FreeTexture(entry);
+			// If at least one entry was not used for the same frame, overwrite the oldest one
+			if (temp_frameCount != 0x7fffffff)
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I think we should increase this check for two frames. So we won't delete textures which may be used later in the current frame just because this frame did use another new one in the beginning.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 9]
```diff
@@ -329,56 +328,92 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+	if (iter != iter_range.second)
 	{
-		// 1. Calculate reference hash:
-		// calculated from RAM texture data for normal textures. Hashes for paletted textures are modified by tlut_hash. 0 for virtual EFB copies.
-		if (g_ActiveConfig.bCopyEFBToTexture && entry->IsEfbCopy())
-			tex_hash = TEXHASH_INVALID;
-
-		// 2. a) For EFB copies, only the hash and the texture address need to match
-		if (entry->IsEfbCopy() && tex_hash == entry->hash && address == entry->addr)
+		if (iter->second->IsEfbCopy())
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: This code says that there is either exactly one efb copy or several common textures within the cache. Is the behavior intended? I'd suggest to move this
if (iter->second->IsEfbCopy() && (g_ActiveConfig.bCopyEFBToTexture || tex_hash == iter->second->hash))
return ReturnEntry(stage, iter->second);
check into the while loop. So also the freeing in line 410 will work fine for efb copys.

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 8 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Hmm, yeah, if there's one new texture, the worst case is that the whole cache gets shifted by 1, and no entry is cached, because it gets overwritten by the texture just before that. But, since the multimap is sorted in the order the elements got added, i think it's likely to replace the changed texture. That is, if the game keeps the order in which the textures are loaded.

On the other hand, if it's not restricted to the current frame, wouldn't the cache blow up? Up to 200% of its original size, if every texture changes?

Hmm, maybe readd the counter, and allow this blow up, if there's more than 1 element already? Also, to check if a texture was used during the last frame, i need to access the global varibale frameCount. Is that really a good idea?

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 9 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: As long as there's only multiple entries for textures during the same frame, that would work. If i implement the other suggestion, it would allow creating multiple entries for efb copies, which i'm strictly against. The palette converter will probably change that, but that's not part of this PR.

Or maybe it wouldn't work. Isn't framCount == FRAMECOUNT_INVALID for efb copies? So if the hash doesn't match, the code would assume it's a 2nd texture for the same frame.

Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Ok, so you can still move it into the loop but also moving the FreeTexture part. I think it will clean up the code a bit.

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 8 above)
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: All of this framecount handling is a bad idea, but I have no better one. But imo this is out of scope for this PR ;)

**Comment by FioraAeterna** [comment]
If we're worried about cache size, is there a reasonable way to purge entries that haven't been used in a while?

**Comment by mimimi085181** [comment]
@FioraAeterna: Look at TextureCache::Cleanup. Every texture, which hasn't been used for TEXTURE_KILL_THRESHOLD frames, gets removed from the cache. On master that's 200 frames, in this PR it is 10. To me that sounds reasonable. Did you mean something else?

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 9 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I have that right now on my computer. It does look cleaner. But i think the logic isn't. For one, IsEfbCopy is checked for every entry in the cache, as opposed to just once.

Then, iter->second is accessed after the entry was removed, which is bad style in my eyes. So, i would add an else statement again, which makes the code look as unclean as before. Or i could use a break? I assume we want to avoid that wherever possible.

And lastly, it might happen in the future that the code changes, and then efb copies are freed 2 times. Efb copies rarely have a hash mismatch, if it all. But should the code change, and then some rare game crashes because of this, nobody except some players will notice.

In general, i like giving efb copies a special treatment, because it makes it more clear that they are not to be handled the same way as the other textures. Well, maybe i overlooked something obvious(again), and there's a good way to do this. If so, please tell me.

**Comment by FioraAeterna** [comment]
That sounds totally reasonable; if that's the case I don't think there's much to worry about in terms of GPU memory usage.

**Comment by mimimi085181** [comment]
Rebased again and moved the IsEfbCopy part into the loop as per degasus' suggestion.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 10]
```diff
@@ -329,56 +328,88 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
+			{
+				// TODO: Print a warning if the format changes! In this case,
+				// we could reinterpret the internal texture object data to the new pixel format
+				// (similar to what is already being done in Renderer::ReinterpretPixelFormat())
+				// TODO: Convert paletted textures, which are efb copies, using the right palette, so they display correctly
+				return ReturnEntry(stage, iter->second);
+			}
+			else
+			{
+				// Keeping an unused efb copy in the cache is pointless, because a new one will be created in CopyRenderTargetToTexture
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: It's not unused, it's changed

**Comment by degasus** [comment]
@dolphin-emu-bot rebuild

**Comment by degasus** [comment]
LGTM
